lib/cc/resolv_override.rb

```diff
@@ -0,0 +1,46 @@
+require "resolv-replace"
+
+module CC
+  # Allows for control of how DNS is resolved so that when we resolve a host for
+  # the purposes of checking that it's not internal, we can ensure that future
+  # resolutions (done as part of making the request) will find the same IP
+  # address.
+  module ResolvOverride
+    def self.get_and_fix_address(host)
+      dns.getaddress(host).tap do |address|
+        fixed.add_address(host, address)
+      end
+    end
+
+    def self.dns
+      @dns ||= Resolv::DNS.new
+    end
+
+    def self.fixed
+      @fixed ||= ResolvOverride::FixedAddress.new
+    end
+
+    class FixedAddress < Resolv::DNS
+      def add_address(host, address)
+        addresses[host] = address
+      end
+
+      def each_address(name, &proc)
+        if addresses.key?(name)
+          yield addresses.fetch(name)
+        end
+      end
+
+      private
+
+      def addresses
+        @addresses ||= {}
+      end
+    end
+  end
+end
+
+Resolv::DefaultResolver.replace_resolvers([
```

Comment:
Intentional. This seems to be the simplest way to achieve what I wanted, do you have any concrete downsides or suggested alternative approaches to where/how to actually trigger the `replace_`?